README.md
# kalendar: ``datetime.date`` extension for climate date measurements

[![license](https://img.shields.io/github/license/OCHA-DAP/kalendar.svg)](https://github.com/OCHA-DAP/kalendar/blob/main/LICENSE)
[![Test Status](https://github.com/OCHA-DAP/kalendar/workflows/tests/badge.svg)](https://github.com/OCHA-DAP/kalendar/actions?query=workflow%3Atests)
[![PyPI Status](https://github.com/OCHA-DAP/kalendar/workflows/PyPI/badge.svg)](https://github.com/OCHA-DAP/kalendar/actions?query=workflow%3APyPI)
[![Documentation Status](https://readthedocs.org/projects/kalendar/badge/?version=latest)](https://kalendar.readthedocs.io/en/latest/?badge=latest)
[![Coverage Status](https://codecov.io/gh/OCHA-DAP/kalendar/branch/main/graph/badge.svg?token=JpWZc5js4y)](https://codecov.io/gh/OCHA-DAP/kalendar)
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit)](https://github.com/pre-commit/pre-commit)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

In climatology, data is often reported in multi-day formats, such
as dekads (a 10 day definition) or pentads (a 5 day definition).
kalendar is a Python package with ``datetime.date`` extensions
for dekadal and pentadal calendar definitions that make
it easier to work with such climate data.

## Installing

Install and update using [pip](https://pip.pypa.io/en/stable/getting-started/):

```shell
pip install kalendar
```

## Usage

The dekadal and pentadal classes are simple subclasses with convenient
constructors for creating the class from ``datetime`` objects (and
inherited constructors from ``datetime.date``), methods for addition
and subtraction, and converting back to ``datetime.date`` objects.

```python
from kalendar import Dekad, Pentad
d = Dekad(2022, 1)
p = Pentad(2013, 51)
```

See the [documentation](https://kalendar.readthedocs.io/en/latest/) for
full examples.

## Contributing

For guidance on setting up a development environment, see the
[contributing guidelines](https://github.com/OCHA-DAP/kalendar/blob/main/CONTRIBUTING.rst)

## Links

- [Documentation](https://kalendar.readthedocs.io/en/latest/)
- [Changes](https://github.com/OCHA-DAP/kalendar/blob/main/CHANGELOG.rst)
- [PyPI Releases](https://pypi.org/project/kalendar/)
- [Source Code](https://github.com/OCHA-DAP/kalendar)
- [Issue Tracker](https://github.com/OCHA-DAP/kalendar/issues)


CHANGELOG.rst
Changelog
=========

All notable changes to ``kalendar`` will be documented in this file.

The format is based on `Keep a
Changelog <https://keepachangelog.com/en/1.0.0/>`__, and this project
adheres to `Semantic
Versioning <https://semver.org/spec/v2.0.0.html>`__.

0.1.1
-----

Fixed
~~~~~

- Addition now correctly calculates the new year on the boundary for dekad and pentad

0.1.0
-----

Added
~~~~~

- Pentad class

Fixed
~~~~~

- ``CONTRIBUTING.rst`` broken links repaired

Changed
~~~~~~~

- Use fixed `requirements.txt` file for development

0.0.2
-----

Removed
~~~~~~~

- Python 3.7 support

Changed
~~~~~~~

- Directory structure to use ``src`` folder
- Fixed testing to include coverage
- Published to PyPi

0.0.1
-----

Added
~~~~~

-  Dekad module
-  Package setup structure
-  Unit tests
-  Pre-commit hooks: black, flake8, mypy plus others
-  Sphinx documentation


tests/test_dekad.py
"""Test dekad module."""

import pickle
from datetime import date, datetime, timedelta

import pytest

from kalendar import Dekad


@pytest.fixture
def dekad():
    """Dekad fixture."""
    return Dekad(year=2022, dekad=1)


def test_addition(dekad):
    """Test addition magic method."""
    assert dekad + 1 == Dekad(2022, 2)
    assert dekad + 35 == Dekad(2022, 36)
    assert dekad + 38 == Dekad(2023, 3)
    assert dekad + timedelta(days=1) == date(year=2022, month=1, day=2)


def test_subtraction(dekad):
    """Test subtraction magic method."""
    assert dekad - 1 == Dekad(2021, 36)
    assert dekad - 38 == Dekad(2020, 35)
    assert dekad - Dekad(2021, 34) == 3
    assert dekad - date(year=2021, month=12, day=20) == timedelta(days=12)


def test_repr(dekad):
    """Test repr magic method."""
    assert eval(repr(dekad)) == dekad


def test_pickling(dekad):
    """Test that pickling works on the class.

    Magic methods sometimes need to be adjusted for
    subclassing or pickling will generate errors.
    This ensures that it's always possible. Based
    on the article that inspired the class.

    https://www.aaronoellis.com/articles/subclassing-datetime-date-in-python-3
    """
    assert pickle.loads(pickle.dumps(dekad)) == dekad


def test_todate(dekad):
    """Test the todate() method."""
    assert dekad.todate() == date(year=2022, month=1, day=1)
    d = Dekad(year=2008, dekad=32)
    assert d.todate() == date(year=2008, month=11, day=11)


def test_constructors(dekad):
    """Test constructor methods."""
    assert Dekad.fromisocalendar(year=2022, week=1, day=7) == dekad
    assert Dekad.fromisoformat("2022-01-10") == dekad
    assert Dekad.fromordinal(738161) == dekad
    assert Dekad.fromdate(date(year=2022, month=1, day=5))
    # datetime
    dt = datetime(year=2022, month=1, day=8)
    assert Dekad.fromtimestamp(dt.timestamp()) == dekad
    assert Dekad.fromdatetime(dt) == dekad


def test_constructors_error():
    """Test constructor error."""
    with pytest.raises(ValueError):
        Dekad(2022, -1)
    with pytest.raises(ValueError):
        Dekad(2011, 37)


def test_printing(dekad, capsys):
    """Test printing."""
    print(dekad)  # noqa: T001
    out, _ = capsys.readouterr()
    assert out == "2022 D1\n"


def test_dekad_month(dekad):
    """Test monthly dekad."""
    assert dekad.dekad_monthly == 1
    assert (dekad + 2).dekad_monthly == 3


tests/test_pentad.py
"""Test pentad module."""

import pickle
from datetime import date, datetime, timedelta

import pytest

from kalendar import Pentad


@pytest.fixture
def pentad():
    """Pentad fixture."""
    return Pentad(year=2022, pentad=1)


def test_addition(pentad):
    """Test addition magic method."""
    assert pentad + 1 == Pentad(2022, 2)
    assert pentad + 72 == Pentad(2022, 73)
    assert pentad + 75 == Pentad(2023, 3)
    assert pentad + timedelta(days=1) == date(year=2022, month=1, day=2)


def test_subtraction(pentad):
    """Test subtraction magic method."""
    assert pentad - 1 == Pentad(2021, 73)
    assert pentad - 75 == Pentad(2020, 72)
    assert pentad - Pentad(2021, 71) == 3
    assert pentad - date(year=2021, month=12, day=20) == timedelta(days=12)


def test_repr(pentad):
    """Test repr magic method."""
    assert eval(repr(pentad)) == pentad


def test_pickling(pentad):
    """Test that pickling works on the class.

    Magic methods sometimes need to be adjusted for
    subclassing or pickling will generate errors.
    This ensures that it's always possible. Based
    on the article that inspired the class.

    https://www.aaronoellis.com/articles/subclassing-datetime-date-in-python-3
    """
    assert pickle.loads(pickle.dumps(pentad)) == pentad


def test_todate(pentad):
    """Test the todate() method."""
    assert pentad.todate() == date(year=2022, month=1, day=1)
    p = Pentad(year=2008, pentad=32)
    assert p.todate() == date(year=2008, month=6, day=5)


def test_constructors(pentad):
    """Test constructor methods."""
    assert Pentad.fromisocalendar(year=2022, week=1, day=3) == pentad
    assert Pentad.fromisoformat("2022-01-03") == pentad
    assert Pentad.fromordinal(738157) == pentad
    assert Pentad.fromdate(date(year=2022, month=1, day=5))
    # datetime
    dt = datetime(year=2022, month=1, day=1)
    assert Pentad.fromtimestamp(dt.timestamp()) == pentad
    assert Pentad.fromdatetime(dt) == pentad


def test_constructors_error():
    """Test constructor error."""
    with pytest.raises(ValueError):
        Pentad(2022, -1)
    with pytest.raises(ValueError):
        Pentad(2011, 75)


def test_printing(pentad, capsys):
    """Test printing."""
    print(pentad)  # noqa: T001
    out, _ = capsys.readouterr()
    assert out == "2022 P1\n"


def test_leapyears():
    """Test dates during leap years."""
    p = Pentad(2012, 12)
    assert p.todate() == date(year=2012, month=2, day=25)
    assert (p + 1).todate() == date(year=2012, month=3, day=2)
    assert Pentad.fromisoformat("2012-02-29") == p
    assert Pentad.fromisoformat("2012-12-27").pentad == 73


src/kalendar/dekad.py
"""Dekad extension for ``datetime.date``."""

from __future__ import annotations

import datetime
from typing import Union

# define Dekad class based off of this article
# https://www.aaronoellis.com/articles/subclassing-datetime-date-in-python-3


class Dekad(datetime.date):
    """Dekad extension for ``datetime.date``.

    Extension of the ``datetime.date`` base class for dekads.
    Uses standard yearly definition of dekad where each month
    is composed of 3 dekads, the first 2 comprising the first 1-10,
    and 11-20 days respectively, and the third dekad comprising
    all remaining days of the month. Under the hood, each ``Dekad``
    is stored as the first day of the dekad, so the 1st dekad of a
    month is the 1st, the 2nd is on the 11th, and the 3rd is on the
    21st. Extension of ``datetime.date``
    allows for comparison of dekads and conversion between
    classes and use of existing constructors, methods, and properties,
    with some additions detailed below. For the purposes of the class
    structure, the yearly definition is relied on, but the monthly
    dekad (1st to 3rd) can be accessed using ``dekad_monthly``.

    Base constructor takes year and dekad. Can
    also be constructed directly from ``datetime.date``
    or ``datetime.datetime`` objects and use any
    ``datetime.date`` constructors:

    ``fromdate()``
    ``fromdatetime()``
    ``fromisoformat()``
    ``fromisocalendar()``
    ``fromtimestamp()``
    ``fromordinal()``

    Method to return ``datetime.date`` object
    added, ``todate()``. Other methods remain
    available. Read-only properties ``dekad`` and
    ``dekad_monthly`` added
    alongside year, month, and day.

    Parameters
    ----------
    year : int
        Year.
    dekad : int
        Dekad, from 1 to 36.

    Examples
    --------
    >>> d = Dekad(2022, 1)
    >>> d
    Dekad(2022, 1)
    >>> d.todate()
    datetime.date(2022, 1, 1)
    >>> d - 1
    Dekad(2021, 36)
    >>> d.dekad
    1
    """

    def __new__(cls, year: int, dekad: int):
        """Initialize new ``Dekad``."""
        if dekad < 1 or dekad > 36:
            raise ValueError("Dekad must be between 1 and 36.")
        month = ((dekad - 1) // 3) + 1
        day = 10 * ((dekad - 1) % 3) + 1  # first day of dekad
        return super().__new__(cls, year, month, day)  # noqa: FKA01

    def __repr__(self):
        """Represent ``Dekad``."""
        return f"{self.__class__.__name__}({self.year}, {self.dekad})"

    def __reduce__(self):
        """Reduce."""
        return (self.__class__, (self.year, self.dekad))

    def __str__(self):
        """Print representation."""
        return f"{self.year} D{self.dekad}"

    def __add__(  # type: ignore[override]
        self, other: Union[int, datetime.timedelta]
    ) -> Union[Dekad, datetime.date]:
        """Addition magic method for dekads.

        For integers, adds that to the dekads and returns a new
        ``Dekad`` object. Otherwise, follows the behavior of
        ``datetime.date``.

        Examples
        --------
        >>> d = Dekad(2021, 36)
        >>> d + 1
        Dekad(2022, 1)

        Returns
        -------
        Union[Dekad, datetime.date]
            Returns ``Dekad`` if adding integer, or
            ``datetime.date`` otherwise.
        """
        if isinstance(other, int):
            new_year = self.year + (self.dekad + other - 1) // 36
            new_dekad = self._dekad_adjuster(self.dekad, other)
            return Dekad(year=new_year, dekad=new_dekad)
        try:
            return self.todate() + other
        except TypeError as e:
            raise TypeError(
                f"unsupported operand type(s) for +: "
                f"'Dekad' and {type(other).__name__}"
            ) from e

    def __sub__(  # type: ignore[override]
        self, other: Union[int, Dekad, datetime.timedelta]
    ) -> Union[Dekad, int, datetime.timedelta]:
        """Subtraction magic method for dekads.

        For integers, subtracts that from the dekads.
        Dekads can also be subtracted from each other,
        returning the number of dekads between. Otherwise,
        follows the behavior of ``datetime.date``.

        Examples
        --------
        >>> import datetime
        >>>
        >>> d1 = Dekad(2022, 1)
        >>> d2 = Dekad(2021, 1)
        >>> d1 - 1
        Dekad(2021, 36)
        >>> d1 - d2
        36
        >>> d1 - datetime.date(2021, 12, 30)
        datetime.timedelta(days=2)

        Returns
        -------
        Union[Dekad, int, datetime.timedelta]
            Returns ``Dekad`` if subtracting integer, or
            ``int`` if subtracting another ``Dekad``,
            and ``datetime.timedelta`` otherwise.
        """
        if isinstance(other, int):
            new_dekad = self._dekad_adjuster(self.dekad, -other)
            new_year = self.year - (36 - self.dekad + other) // 36
            return Dekad(year=new_year, dekad=new_dekad)
        if isinstance(other, Dekad):
            return self.dekad - other.dekad + 36 * (self.year - other.year)
        try:
            return self.todate() - other  # type: ignore
        except TypeError as e:
            raise TypeError(
                f"unsupported operand type(s) for -: "
                f"'Dekad' and {type(other).__name__}"
            ) from e

    @property
    def dekad(self) -> int:
        """Dekad of the year, 1 to 36."""
        return self._get_dekad(month=self.month, day=self.day)

    @property
    def dekad_monthly(self) -> int:
        """Dekad of the month, 1 to 3."""
        return 1 + (self.day - 1) // 10

    def todate(self) -> datetime.date:
        """Convert to datetime.date object for the year, month, and day."""
        return datetime.date(year=self.year, month=self.month, day=self.day)

    @classmethod
    def fromdate(cls, d: datetime.date) -> Dekad:
        """Construct a dekad from a datetime.date object."""
        dekad = cls._get_dekad(month=d.month, day=d.day)
        return cls(year=d.year, dekad=dekad)

    @classmethod
    def fromdatetime(cls, dt: datetime.datetime) -> Dekad:
        """Construct a dekad from a datetime.date object."""
        dekad = cls._get_dekad(month=dt.month, day=dt.day)
        return cls(year=dt.year, dekad=dekad)

    @classmethod
    def fromisoformat(cls, time_string: str) -> Dekad:
        """Construct a dekad from a string in one of the ISO 8601 formats."""
        return cls.fromdate(datetime.date.fromisoformat(time_string))

    @classmethod
    def fromisocalendar(cls, year: int, week: int, day: int) -> Dekad:
        """Construct a dekad from the ISO year, week number and weekday."""
        return cls.fromdate(
            datetime.date.fromisocalendar(year=year, week=week, day=day)
        )

    @classmethod
    def fromtimestamp(cls, timestamp: float) -> Dekad:
        """Construct a dekad from a POSIX timestamp (like time.time())."""
        return cls.fromdate(datetime.date.fromtimestamp(timestamp))

    @classmethod
    def fromordinal(cls, n: int) -> Dekad:
        """Construct a dekad from a proleptic Gregorian ordinal.

        January 1 of the year 1 is day 1. Only the year, month and day
        are non-zero in the result.
        """
        return cls.fromdate(datetime.date.fromordinal(n))

    @staticmethod
    def _dekad_adjuster(d1: int, d2: int) -> int:
        """Add or subtract d2 from d1 and keep within dekadal ranges."""
        return (d1 + (d2 % 36) + 35) % 36 + 1

    @staticmethod
    def _get_dekad(month: int, day: int) -> int:
        """Get dekad from day and month."""
        return min((day - 1) // 10, 2) + ((month - 1) * 3) + 1


src/kalendar/pentad.py
"""Pentad extension for ``datetime.date``."""

from __future__ import annotations

import calendar
import datetime
from typing import Union

# define Pentad class based off of this article
# https://www.aaronoellis.com/articles/subclassing-datetime-date-in-python-3


class Pentad(datetime.date):
    """Pentad extension for ``datetime.date``.

    Extension of the ``datetime.date`` base class for pentads.
    Uses standard yearly definition of pentads where the 365 day
    calendar year is divided into 73 pentads of 5 days each.
    Under the hood, each ``Pentad``
    is stored as the first day of the pentad, the 1st pentad of the year
    is on January 1st, the 2nd on January 6th, and so on. Leap years have
    366 days, and pentad 12 is extended comprise 6 days that
    includes February 29th. Extension of ``datetime.date``
    allows for comparison of pentads and conversion between
    classes and use of existing constructors, methods, and properties,
    with some additions detailed below.

    Base constructor takes year and pentad. Can
    also be constructed directly from ``datetime.date``
    or ``datetime.datetime`` objects and use any
    ``datetime.date`` constructors:

    ``fromdate()``
    ``fromdatetime()``
    ``fromisoformat()``
    ``fromisocalendar()``
    ``fromtimestamp()``
    ``fromordinal()``

    Method to return ``datetime.date`` object
    added, ``todate()``. Other methods remain
    available. Read-only property ``pentad``
    added alongside year, month, and day.

    Parameters
    ----------
    year : int
        Year.
    pentad : int
        Pentad, from 1 to 73.

    Examples
    --------
    >>> p = Pentad(2022, 1)
    >>> p
    Pentad(2022, 1)
    >>> p.todate()
    datetime.date(2022, 1, 1)
    >>> p - 1
    Pentad(2021, 73)
    >>> p.pentad
    1
    """

    def __new__(cls, year: int, pentad: int):
        """Initialize new ``Pentad``."""
        if pentad < 1 or pentad > 73:
            raise ValueError("Pentad must be between 1 and 73.")

        # get pentad month and day from non-leap year
        dt = datetime.date(year=2010, month=1, day=1) + datetime.timedelta(
            (pentad - 1) * 5
        )

        return super().__new__(cls, year, dt.month, dt.day)  # noqa: FKA01

    def __repr__(self):
        """Represent ``Pentad``."""
        return f"{self.__class__.__name__}({self.year}, {self.pentad})"

    def __reduce__(self):
        """Reduce."""
        return (self.__class__, (self.year, self.pentad))

    def __str__(self):
        """Print representation."""
        return f"{self.year} P{self.pentad}"

    def __add__(  # type: ignore[override]
        self, other: Union[int, datetime.timedelta]
    ) -> Union[Pentad, datetime.date]:
        """Addition magic method for pentads.

        For integers, adds that to the pentads and returns a new
        ``Pentad`` object. Otherwise, follows the behavior of
        ``datetime.date``.

        Examples
        --------
        >>> p = Pentad(2021, 73)
        >>> p + 1
        Pentad(2022, 1)

        Returns
        -------
        Union[Pentad, datetime.date]
            Returns ``Pentad`` if adding integer, or
            ``datetime.date`` otherwise.
        """
        if isinstance(other, int):
            new_year = self.year + (self.pentad + other - 1) // 73
            new_pentad = self._pentad_adjuster(self.pentad, other)
            return Pentad(year=new_year, pentad=new_pentad)  # type: ignore
        try:
            return self.todate() + other
        except TypeError as e:
            raise TypeError(
                f"unsupported operand type(s) for +: "
                f"'Pentad' and {type(other).__name__}"
            ) from e

    def __sub__(  # type: ignore[override]
        self, other: Union[int, Pentad, datetime.timedelta]
    ) -> Union[Pentad, int, datetime.timedelta]:
        """Subtraction magic method for pentads.

        For integers, subtracts that from the pentads.
        Pentads can also be subtracted from each other,
        returning the number of pentads between. Otherwise,
        follows the behavior of ``datetime.date``.

        Examples
        --------
        >>> import datetime
        >>>
        >>> p1 = Pentad(2022, 1)
        >>> p2 = Pentad(2021, 1)
        >>> p1 - 1
        Pentad(2021, 73)
        >>> p1 - p2
        73
        >>> p1 - datetime.date(2021, 12, 30)
        datetime.timedelta(days=2)

        Returns
        -------
        Union[Pentad, int, datetime.timedelta]
            Returns ``Pentad`` if subtracting integer, or
            ``int`` if subtracting another ``Pentad``,
            and ``datetime.timedelta`` otherwise.
        """
        if isinstance(other, int):
            new_pentad = self._pentad_adjuster(self.pentad, -other)
            new_year = self.year - (73 - self.pentad + other) // 73
            return Pentad(year=new_year, pentad=new_pentad)
        if isinstance(other, Pentad):
            pentad_diff = self.pentad - other.pentad
            year_diff = 73 * (self.year - other.year)
            return pentad_diff + year_diff  # type: ignore
        try:
            return self.todate() - other  # type: ignore
        except TypeError as e:
            raise TypeError(
                f"unsupported operand type(s) for -: "
                f"'Pentad' and {type(other).__name__}"
            ) from e

    @property
    def pentad(self) -> int:
        """Pentad of the year, 1 to 73."""
        return self._get_pentad(dt=self.todate())

    def todate(self) -> datetime.date:
        """Convert to datetime.date object for the year, month, and day."""
        return datetime.date(year=self.year, month=self.month, day=self.day)

    @classmethod
    def fromdate(cls, d: datetime.date) -> Pentad:
        """Construct a pentad from a datetime.date object."""
        pentad = cls._get_pentad(dt=d)
        return cls(year=d.year, pentad=pentad)

    @classmethod
    def fromdatetime(cls, dt: datetime.datetime) -> Pentad:
        """Construct a pentad from a datetime.date object."""
        pentad = cls._get_pentad(dt=dt)
        return cls(year=dt.year, pentad=pentad)

    @classmethod
    def fromisoformat(cls, time_string: str) -> Pentad:
        """Construct a pentad from a string in one of the ISO 8601 formats."""
        return cls.fromdate(datetime.date.fromisoformat(time_string))

    @classmethod
    def fromisocalendar(cls, year: int, week: int, day: int) -> Pentad:
        """Construct a pentad from the ISO year, week number and weekday."""
        return cls.fromdate(
            datetime.date.fromisocalendar(year=year, week=week, day=day)
        )

    @classmethod
    def fromtimestamp(cls, timestamp: float) -> Pentad:
        """Construct a pentad from a POSIX timestamp (like time.time())."""
        return cls.fromdate(datetime.date.fromtimestamp(timestamp))

    @classmethod
    def fromordinal(cls, n: int) -> Pentad:
        """Construct a pentad from a proleptic Gregorian ordinal.

        January 1 of the year 1 is day 1. Only the year, month and day
        are non-zero in the result.
        """
        return cls.fromdate(datetime.date.fromordinal(n))

    @staticmethod
    def _pentad_adjuster(p1: int, p2: int) -> int:
        """Add or subtract p2 from p1 and keep within pentadal ranges."""
        return (p1 + (p2 % 73) + 72) % 73 + 1

    @staticmethod
    def _get_pentad(dt: datetime.date) -> int:
        """Get pentad from datetime."""
        yday = dt.timetuple().tm_yday
        if calendar.isleap(dt.year) and yday >= 60:
            return 1 + (yday - 2) // 5
        else:
            return 1 + (yday - 1) // 5
